source/tutorials/file-sets.md

```diff
@@ -0,0 +1,725 @@
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+this derivation will be built:
+  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
+...
+'hello.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/hello.txt'
+'world.txt' -> '/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset/world.txt'
+...
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+An attempt to repeat the build will re-use the existing store path:
+
+```
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+## Missing files
+
+Removing the `./result` symlink creates a new problem, though:
+
+```shell-session
+$ rm result
+$ nix-build
+error: lib.fileset.difference: Second argument (negative set)
+  (/home/user/select/result) is a path that does not exist.
+  To create a file set from a path that may not exist, use `lib.fileset.maybeMissing`.
+```
+
+Follow the instructions in the error message, and use `maybeMissing` to create a file set from a path that may not exist (in which case the file set will be empty):
+
+<!-- https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.maybeMissing -->
+
+```{code-block} diff
+:caption: build.nix
+ { stdenv, lib }:
+ let
+   fs = lib.fileset;
+-  sourceFiles = fs.difference ./. ./result;
++  sourceFiles = fs.difference ./. (fs.maybeMissing ./result);
+ in
+```
+
+This now works, using the whole directory since `./result` is not present:
+
+```
+$ nix-build
+trace: /home/user/select (all files in directory)
+this derivation will be built:
+  /nix/store/zr19bv51085zz005yk7pw4s9sglmafvn-fileset.drv
+...
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+Another build attempt will produce a different trace, but the same output path:
+
+```
+$ nix-build
+trace: /home/user/fileset
+trace: - build.nix (regular)
+trace: - default.nix (regular)
+trace: - hello.txt (regular)
+trace: - nix (all files in directory)
+trace: - world.txt (regular)
+/nix/store/vhyhk6ij39gjapqavz1j1x3zbiy3qc1a-fileset
+```
+
+## Union (explicitly exclude files)
+
+There is still a problem:
+Changing _any_ of the included files causes the derivation to be rebuilt, even though it doesn't depend on those files.
+
+Append an empty line to `build.nix`:
+
+```shell-session
+$ echo >> build.nix
+```
+
+Again, Nix will start from scratch:
+
+```shell-session
+$ nix-build
+trace: /home/user/select
+trace: - default.nix (regular)
+trace: - nix (all files in directory)
+trace: - build.nix (regular)
+trace: - string.txt (regular)
+this derivation will be built:
+  /nix/store/zmgpqlpfz2jq0w9rdacsnpx8ni4n77cn-filesets.drv
+...
+/nix/store/6pffjljjy3c7kla60nljk3fad4q4kkzn-filesets
+```
+
+One way to fix this is to use [`unions`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.unions).
```

Comment:
```
One way to fix this is with [`unions`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.unions).
```

but wait, we also need difference?
Maybe start with `unions` without difference? That's actually a great way to make sure you're not including too much. Nix's own flake does that.

Then later flip it around: if you prefer to include any new or previously unknown files by default, you can use `unions` in a `difference` instead.

Reply:
I found it hard to make the tutorial flow that way, but I can take another look

Reply:
See https://github.com/NixOS/nix.dev/pull/802#discussion_r1402328626